### PR TITLE
Aseprite tilemap support

### PR DESCRIPTION
**Pull request by soywiz**
TileMap/TileSet functionality moved here: https://github.com/korlibs/korge/commit/28f37ce42065757566299bd4057f28fa54819583

Implemented the spec diff from here: https://github.com/aseprite/aseprite/commit/95e6b5ddd6eed2ab80f8d2f19109ba4d9931ad5b

Fixes https://github.com/korlibs/korge/issues/504

### CODE REVIEW

**Comment by soywiz** [comment]
Still misses support on KorGE's side

**Comment by soywiz** [comment]
Screenshot 2022-06-30 at 23 05 44

**Comment by soywiz** [comment]
@jobe-m can you have a look? Since the view includes something related to view recycling Im not sure if the proposed implementation would work for you

**Comment by jobe-m** [comment]
Cool, yes, I will have a look!

File: korim/src/commonMain/kotlin/com/soywiz/korim/format/ASE.kt [hunk 1]
```diff
@@ -59,24 +65,54 @@ object ASE : ImageFormatWithContainer("ase") {
...
+        val x: Int
+        val y: Int
+        val opacity: Int
+        fun resolve(): AseCell = this
+    }
+    open class AseBitmapCell(
+        override val bmp: Bitmap,
+        override val x: Int, override val y: Int, override val opacity: Int,
+    ) : AseEntity by AseEntity(), AseCell
+    open class AstLinkedCell(
```
Comment (soywiz) [inline korim/src/commonMain/kotlin/com/soywiz/korim/format/ASE.kt:79]: Typo: ast->ase

**Comment by jobe-m** [comment]
I checked the parallax-scrolling-aseprite example as it uses also classes which you changed here. It look somewhat good as it still shows all layer objects but I guess there is a -1/+1 problem with repeating the TileMap (IIRC) objects. Every first object from the left is not shown and it starts to appear when the full object is visible on the screen while moving to the right side. See picture:

image

File: korge/src/commonMain/kotlin/com/soywiz/korge/view/animation/ImageAnimationView.kt [hunk 2]
```diff
@@ -111,15 +131,26 @@ open class ImageAnimationView<T: SmoothedBmpSlice>(
...
         nframes = animation?.frames?.size ?: 1
         // Before clearing layers let parent possibly recycle layer objects (e.g. return to pool, etc.)
-        for (layer in layers) { onDestroyLayer?.invoke(layer) }
+        for (layer in layers) {
+            if (layer is TileMap) {
+                onDestroyTilemapLayer?.invoke(layer)
+            } else {
+                onDestroyLayer?.invoke(layer as T)
+            }
+        }
```
Comment (jobe-m) [inline korge/src/commonMain/kotlin/com/soywiz/korge/view/animation/ImageAnimationView.kt:140]: LGTM!

File: korge-sandbox/src/commonMain/kotlin/Main.kt [hunk 3]
```diff
@@ -94,6 +94,7 @@ suspend fun main() = Korge(
             Demo(::MainTilemapWithScroll),
             Demo(::MainTiledBackground),
             Demo(::MainAseprite),
+            Demo(::MainAseParallaxSample),
```
Comment (jobe-m) [inline korge-sandbox/src/commonMain/kotlin/Main.kt:97]: 👍

**Comment by soywiz** [comment]
So the parallax sample seems to be broken, but doesn't seems to be related to this right? I believe it was broken when optimizations regarding visibily were done (for the other sample, the one that displays a huge tilemap with a donut), I have tried manually adjusting min/maxXY in the tilemap rendering and seems to paint stuff.
I have open a separate issue to investigate that: https://github.com/korlibs/korge/issues/512
Maybe we should also consider refactoring it, because it is getting a bit complicated

**Comment by soywiz** [comment]
BTW the `SingleTile` view is only used to repeat an image in the parallax code, or are you using it somewhere else? If I refactor tilemaps and parallax to not require that class, would be fine for you?

**Pull request merged**